test/test_action_builder/test_action_builder.py

```diff
@@ -0,0 +1,345 @@
+        )
+
+    @requests_mock.Mocker()
+    def test_get_entry_list(self, m):
+        m.get(
+            f"{self.api_url}/tags?page=1&per_page=25",
+            text=json.dumps(self.fake_tags_list_1),
+        )
+        m.get(
+            f"{self.api_url}/tags?page=2&per_page=25",
+            text=json.dumps(self.fake_tags_list_2),
+        )
+        m.get(
+            f"{self.api_url}/tags?page=3&per_page=25",
+            text=json.dumps({"_embedded": {"osdi:tags": []}}),
+        )
+        assert_matching_tables(
+            self.bldr._get_entry_list(self.campaign, "tags"), Table(self.fake_tags_list)
+        )
+
+    @requests_mock.Mocker()
+    def test_get_campaign_tags(self, m):
+        m.get(
+            f"{self.api_url}/tags?page=1&per_page=25",
+            text=json.dumps(self.fake_tags_list_1),
+        )
+        m.get(
+            f"{self.api_url}/tags?page=2&per_page=25",
+            text=json.dumps(self.fake_tags_list_2),
+        )
+        m.get(
+            f"{self.api_url}/tags?page=3&per_page=25",
+            text=json.dumps({"_embedded": {"osdi:tags": []}}),
+        )
+        assert_matching_tables(
+            self.bldr.get_campaign_tags(), Table(self.fake_tags_list)
+        )
+
+    @requests_mock.Mocker()
+    def test_get_tag_by_name(self, m):
+        m.get(
+            f"{self.api_url}/tags?filter=name eq '{self.fake_tag_1}'",
+            text=json.dumps(self.fake_tag_name_search_result),
+        )
+        m.get(
+            f"{self.api_url}/tags?page=2&per_page=25&filter=name eq '{self.fake_tag_1}'",
+            text=json.dumps({"_embedded": {"osdi:tags": []}}),
+        )
+        assert_matching_tables(
+            self.bldr.get_tag_by_name(self.fake_tag_1),
+            Table([self.fake_tags_list_1["_embedded"]["osdi:tags"][0]]),
+        )
+
+    def compare_to_incoming(self, dict1, dict2):
+        # Internal method to compare a reference dict to a new incoming one, keeping only common
+        # keys (or where incoming key is substring). Recursive to handle nesting/JSON.
+
+        incoming = {[key for key in dict1 if k in key][0]: v for k, v in dict2.items()}
+        compare = {k: v for k, v in dict1.items() if k in incoming}
+
+        for key, val in incoming.items():
+            print(key)
+            if isinstance(val, dict):
+                insub, compsub = self.compare_to_incoming(compare[key], val)
+                incoming[key] = insub
+                compare[key] = compsub
+            elif isinstance(val, list) and isinstance(val[0], dict):
+                for i in range(len(val)):
+                    print(i)
+                    insub, compsub = self.compare_to_incoming(compare[key][i], val[i])
+                    incoming[key][i] = insub
+                    compare[key][i] = compsub
+
+        return incoming, compare
+
+    @requests_mock.Mocker()
+    def test_upsert_entity(self, m):
```

Comment:
there's a lot going on in this test along with the `compare_to_incoming`. I'm not sure I really follow what's happening so might be easier to grab 10 min and talk through it

Reply:
Happy to do that. In a nutshell, though, `compare_to_incoming()` is just trying to make sure to exclude out any keys (including in nested dicts) from an incoming dict that don't also appear on the existing dict to which the incoming dict is to be compared. In other words, this if an incoming dict has extra keys anywhere, don't include them in the comparison.